### DISCUSSION TITLE
Potential problem with `undo_ftplugin`??

### DISCUSSION THREAD

**Discussion by dpetka2001** [ibhagwan/fzf-lua#1555] [Q&A] [answered]
Hi there. A [discussion](https://github.com/LazyVim/LazyVim/discussions/4966) popped up in LazyVim Discussions about some keymaps not available within php files. The OP mentioned that the first buffer had the mappings but subsequent ones didn't have correct mappings. I found out [this in Neovim php ftplugin](https://github.com/neovim/neovim/blob/fac96b72a59e15d483e03bdc43b9ba08ce6c828e/runtime/ftplugin/php.vim#L129-L150), which seems to be the root of the problem. 

However, setting `vim.g.no_php_maps = true`, just preserves the mapping on the first buffer and deletes the ones in subsequent opened buffers. Quoting here my latest observations
> So, @chronon here's some things that I've observed. I use fzf-lua and observed the same thing you did even with vim.g.no_php_maps = true. But when I open 2 php files with nvim file1.php file2.php, then the mappings work fine on both buffers. If I open Neovim and restore the session where both files are open in buffers, then the mappings work as expected. When I open Neovim with one file and then open another one using Neotree, then the mappings work as expected on both buffers. I even tried switching from fzf-lua to telescope and there it also worked well.
So, from my testing I've observed that when opening a file through FzfLua files picker, then for some reason the undo_ftplugin gets triggered and that deletes the mappings in the buffers that are opened through it. I'm not 100% sure that is what happens, but from my observations this is the explanation I've come up with.

@ibhagwan Do you maybe have some insight as to why this might be happening? My deduction (which of course might be wrong) based on my observations, was that for some reason `undo_ftplugin` was triggered for php files and that deleted the mappings from subsequent buffers. The same thing did not happen when opening `nvim file1.php file2.php` or from Neotree or from Telescope or from restoring a session where both buffers happened to be restored and opened.

`vim.g.no_php_maps` seems to be working for other ways opening files but not with `FzfLua files`. Please do correct me whatever wrong observations I've made. Not sure if this is an issue/bug with FzfLua, so wanted to check about your insight in Discussions first.

**Comment by ibhagwan** [comment]
> Not sure if this is an issue/bug with FzfLua, so wanted to check about your insight in Discussions first.

When it comes to these type of issues I don’t rule out anything, if I could reproduce this locally I can look into it, is installing the mentioned LSP and opening a php enough to reproduce this?

**Reply by dpetka2001** [reply]
Yes i just installed inteliphense via Mason and opened 2 php files generated by gpt (since i don't do any php at all) just to test out the behavior. Lazyvim has `]]/[[` mapped for next/previous reference by default.

**Reply by dpetka2001** [reply]
Also from OP this seems to happen for only php buffers. If you open first php and then another buffer with different filetype and LSP server it doesn't happen. Seems to be related to php ftplugin `undo_ftplugin` kicking in for some reason. No idea why.

**Reply by ibhagwan** [reply]
I read the full discussion again as well as the ftplugin code you linked, that is one weird issue, all fzf-lua does in regards to the file is basically a `:lua vim.cmd('e <file>')` which I would assume then trigger the ftplugin, which in turn deletes the mappings.

Can you try opening the 2nd buffer with `:lua vim.cmd('e <file>')` instead of fzf-lua and see if the problem persists?

**Reply by dpetka2001** [reply]
No, directly calling `:lua vim.cmd("e test2.php")` does not trigger the issue.

**Reply by ibhagwan** [reply]
> No, directly calling `:lua vim.cmd("e test2.php")` does not trigger the issue.

Perhaps it’s some combination of opening a terminal that triggers this? What happens if you first run `:new | term` and then press `<C-\><C-n>` and `:lua vim.cmd("e test2.php")`

This is obviously related to fzf-lua but unlikely a direct cause, it’s some sort of neovim ft runtime bug triggered by the conditions set by fzf-lua.

**Reply by dpetka2001** [reply]
No, that too works ok. Another thing I noticed is that when doing `FzfLua files` and then press `<alt-q>` to send selected file to qflist and then open the file from qflist, then that still works fine. So, to me it seems that it's related to the action of opening the file from the picker. Of course this is purely based on my observations of trying things out and in no aspect meant to be technical.

**Reply by ibhagwan** [reply]
> No, that too works ok. Another thing I noticed is that when doing `FzfLua files` and then press `<alt-q>` to send selected file to qflist and then open the file from qflist, then that still works fine. So, to me it seems that it's related to the action of opening the file from the picker. Of course this is purely based on my observations of trying things out and in no aspect meant to be technical.

It’s a reasonable conclusion, looking at the file open code I can only thing this is triggered by one of the option testing using `vim.o.xxx`, can you try adding this to your setup and opening a file:
```lua
local fzf = require('fzf-lua')
fzf.setup({
  actions = { files = {
    enter = function(sel,opts)
      local entry = path.entry_to_file(sel[1], opts, opts._uri)
      vim.cmd("e " .. entry.path)
    end
  }}
})
```

**Reply by dpetka2001** [reply]
I tried the following which should be equivalent to what you proposed (it also overrides LazyVim's default settings for FzfLua I believe)
```lua
return {
  {
    "ibhagwan/fzf-lua",
    config = function()
      local fzf = require("fzf-lua")
      local path = require("fzf-lua.path")

      fzf.setup({
        actions = {
          files = {
            enter = function(sel, opts)
              local entry = path.entry_to_file(sel[1], opts, opts._uri)
              vim.cmd("e " .. entry.path)
            end,
          },
        },
      })
    end,
  },
}
```
And the result is the same as my previous tests. The mappings get deleted in the second buffer.

**Reply by ibhagwan** [reply]
Maybe it’s something that happens in the entry_to_file code, what if you set the action to a simple one line  `vim.cmd("e test2.php")`?

**Reply by dpetka2001** [reply]
I also tried 
```lua
enter = function(sel, opts)
              vim.cmd("e " .. sel[1])
            end,
```
and it has the same problem. Of course the path doesn't resolve correctly to the file, but it still opens another buffer `test2.php` with empty contents, but the mappings still are absent. So, it seems related to the `actions` itself maybe?

**Reply by dpetka2001** [reply]
Doing a `:LspRestart` returns the mappings. These mappings are defined in LazyVim on the `LspAttach` event and also when LSP dynamic capabilities get resolved (but I don't understand much about dynamic capabilities resolution to be honest).

**Reply by ibhagwan** [reply]
> So, it seems related to the actions itself maybe?

The how comes it works when sending to the quick fix?

**Reply by ibhagwan** [reply]
What if you run with `:FzfLua files previewer=false` or previewer=bat?

**Reply by dpetka2001** [reply]
Even if I do 
```lua
        actions = {
          files = {
            enter = function(sel, opts)
              -- local entry = path.entry_to_file(sel[1], opts, opts._uri)
              vim.cmd("e test2.php")
            end,
          },
        },
```
Mappings still get deleted. When you send it to the quickfix list you don't open directly the file, you just execute an action to send it to quickfix. Quickfixlist handles the opening of the file when you click on the quickfix entry. That would be my guess??

**Reply by dpetka2001** [reply]
Both `previewer = false` and `previewer=bat` have the same outcome, deleted mappings.

**Reply by ibhagwan** [reply]
> Mappings still get deleted. When you send it to the quickfix list you don't open directly the file, you just execute an action to send it to quickfix. Quickfixlist handles the opening of the file when you click on the quickfix entry. That would be my guess??

Perhaps the combination of opening the file from the fzf-lua float which would lead to something about the options of the float win? What happens if you try `:FzfLua files winopts.split=belowright\ new` or use `:FzfLua profiles` and change to fzf-tmux (if you’re using tmux)?

**Reply by dpetka2001** [reply]
`:FzfLua files winopts.split=belowright\ new` still same problem. `fzf-tmux` profile outside of tmux environment behaves the same. However, inside a tmux session `fzf-tmux` profile doesn't have the same problem. When I open second buffer the mappings don't get deleted.

**Reply by ibhagwan** [reply]
> `:FzfLua files winopts.split=belowright\ new` still same problem. `fzf-tmux` profile outside of tmux environment behaves the same. However, inside a tmux session `fzf-tmux` profile doesn't have the same problem. When I open second buffer the mappings don't get deleted.

So it works from tmux (no fzf neovim win), works from quick fix (generated by fzf-lua) but doesn’t work from a neovim terminal window running fzf-lua regardless of float/split.

The above leads me to conclude this is a neovim “bug” related to the terminal / window options, not caused by fzf-lua but related.

**Reply by ibhagwan** [reply]
Maybe try this again with the addition of file type?

- `:new | term`
-  press `<C-\><C-n>`
- run `:set ft=fzf`
- and `:lua vim.cmd("e test2.php")`

**Reply by dpetka2001** [reply]
I don't know what to tell you. You're obviously more knowledgeable, so you have a better understanding as to what might be going on. OP from LazyVim discussions also confirmed that he was using FzfLua and that opening the second buffer via another way (Neotree, Telescope, session) resolved the problem. 

If you would like to open a Neovim issue and give a proper description of what's going on and FzfLua's process under the hood, it would be greatly appreciated, but for sure you're not obligated to do that. I just believe that if you opened the issue, you would be able to have a more meaningful discussion with Neovim devs as to what's going on.

Either way, really appreciate you taking the time to look into this from all possible angles you could think of, because I would definitely not be able to make progress regarding this matter on my own. Thank you again!!

**Reply by dpetka2001** [reply]
> Maybe try this again with the addition of file type?
    :new | term
    press <C-\><C-n>
    run :set ft=fzf
    and :lua vim.cmd("e test2.php")

This works correctly.

**Reply by dpetka2001** [reply]
The last one is similar to what you suggested in [this comment](https://github.com/ibhagwan/fzf-lua/discussions/1555#discussioncomment-11494551) and both work correctly.

**Reply by ibhagwan** [reply]
> I don't know what to tell you. You're obviously more knowledgeable, so you have a better understanding as to what might be going on. OP from LazyVim discussions also confirmed that he was using FzfLua and that opening the second buffer via another way (Neotree, Telescope, session) resolved the problem.

You’re giving me too much credit here, I had no idea about `undo_ftplugin` until I read your post :)

As I mentioned, this is triggered by fzf lua somehow but not within its “responsibility” so to speak.


> If you would like to open a Neovim issue and give a proper description of what's going on and FzfLua's process under the hood, it would be greatly appreciated, but for sure you're not obligated to do that. I just believe that if you opened the issue, you would be able to have a more meaningful discussion with Neovim devs as to what's going on.

I wouldn’t open a neovim issue unless I know the exact reason and conditions for this bug other than “open a file with fzf-lua”.

> Either way, really appreciate you taking the time to look into this from all possible angles you could think of, because I would definitely not be able to make progress regarding this matter on my own. Thank you again!!

I’ll try and think of other ideas to test :)

**Reply by dpetka2001** [reply]
What is the difference between `fzf-tmux` profile and the default one? Doesn't the `fzf-tmux` profile run in a terminal window?

**Reply by ibhagwan** [reply]
> What is the difference between `fzf-tmux` profile and the default one? Doesn't the `fzf-tmux` profile run in a terminal window?

In this case Fzf runs in a tmux pop up as opposed to a neovim window.

**Reply by dpetka2001** [reply]
Just for clarification, but were you able to reproduce this on your end? Or is this specifically tied to LazyVim?

**Reply by ibhagwan** [reply]
> Just for clarification, but were you able to reproduce this on your end? Or is this specifically tied to LazyVim?

I did not get to trying to reproduce this.

**Reply by dpetka2001** [reply]
Either way thank you for your time as always. If you want me to test further things, just say what to do in this discussion and I'll try that and give you back whatever feedback I can get from my observations.

**Comment by ibhagwan** [comment]
I'm trying to reproduce this to see what triggers this behavior, used the project of the user who opened this issue: https://github.com/chronon/cakephp-remote-files

Installed `intelephense` via Mason (you can see it's working in the statusline), I open a php file directly from the shell `nvim src/RemoteFilesPlugin.php` but I don't see the mappings for nex/prev references, also tried `:LspRestart intelephense` yet I only see the "bad" mapping.

What am I missing @dpetka2001 @chronon?

![image](https://github.com/user-attachments/assets/2b679f0f-8903-48d8-892a-a8101d0e3ff0)

**Reply by ibhagwan** [reply]
Apparently the `Next|Prev Reference` mappings are defined by LazyVim's LSP config:
https://github.com/LazyVim/LazyVim/blob/c8d8dd0ab6b4db711b33a79835cbac3b389b37f4/lua/lazyvim/plugins/lsp/keymaps.lua#L32

The ftplugin mappings are defined (for the buffer only) by neovim's runtime which makes total sense to run on every open of a php file, I have no idea why this doesn't get triggered by Telescope/Neotree (IMHO it should or you're missing on important filetype settings, perhaps it's a loading order thing?

In any event there's a simple workaround to this by addding the below to your config:
```lua
vim.g.no_php_maps = 1
```
Which then causes `vim.fn.exists("no_php_maps")` to return `1` (true in vimL).

By setting the above I no longer get these mappings defined, even when opening 2nd+ buffers (as noted in the original issue).

**Reply by ibhagwan** [reply]
And btw, I also tried Telescope, the behavior on my machine is the same on fzf-lua and Telescope on both scenarios:
Scenario **(1)**
  - No `vim.g.no_php_maps` in my config
  - Open a php buffer from the command line
  - `]]` is defined - **correct behavior**
  - Run `:lua vim.g.no_php_maps=1`
  - Open a new buffer from either fzf-lua or Telescope
  - Result **No `]]` mapping, correct behavior**

Scenario **(2)**
  - Have `vim.g.no_php_maps=1` in my config
  - Open php buffer from command line **No `]]` mapping, correct behavior**
  - Open a 2nd buffer from either fzf-lua or Telescope **No `]]` mapping, correct behavior**

Both cases the correct behavior is observed, the `ftplugin/php.vim` is run (as it should) and mappings are either defined or skipped based on the value of `vim.g.no_php_maps`.

**Reply by ibhagwan** [reply]
Btw, @dpetka2001, this has nothing to with `undo_ftpplugin` which is just a store of commands to undo these mappings if the file type changes with `:set ft=…`, these mappings are executed when the runtime file is loaded, here:
https://github.com/neovim/neovim/blob/fac96b72a59e15d483e03bdc43b9ba08ce6c828e/runtime/ftplugin/php.vim#L144-L146
```vim
  for mode in ["n", "o", "x"]
    exe mode .. "noremap <buffer> <silent> ]] <Cmd>call <SID>Jump('" .. s:section .. "', v:count1, 'W')<CR>"
    exe mode .. "noremap <buffer> <silent> [[ <Cmd>call <SID>Jump('" .. s:section .. "', v:count1, 'bW')<CR>"
```

`undo_ftplugin` is then also set which is a command to clear these mappings but issue here remains, that as long as the runtime is sourced (which it should…) the mappings will be set unless you set `no_php_maps`.

**Reply by dpetka2001** [reply]
@ibhagwan I appreciate you taking the time to look into this. But I can still reproduce what I described in my initial post here and in the OP LazyVim discussion. 

You seeing no `]]` mappings in the first buffer has to do with no LSP attaching to the file i believe. What does `:LspInfo` show? If you have `intelephense` attached, then in the first buffer you should see the `]] Next Reference` mapping in `which-key` (it's just a visual confirmation and LazyVim also uses it, nothing to do with this case). For `intelephense` to be attached I just did `git init` as per [lspconfig docs](https://github.com/neovim/nvim-lspconfig/blob/ae8a01bb40848490548e5d07b97ed972ed43c2f9/lua/lspconfig/configs/intelephense.lua#L9) because it seemed the simpler way to me, since I don't do PHP and didn't know if I had to configure correctly a `composer.json` or if I could just leave it empty.

Here'a s screencast with default LazyVim installation showing everything that happens

[Screencast_2024-12-16-11-59-45.webm](https://github.com/user-attachments/assets/21356c55-a37c-4fcb-ab70-8dcdf38635e1)

PS: I understand about `undo_ftplugin` and when the mappings are created. For some reason in the second buffer the LSP mappings which get created via a `LspAttach` autocmd are not there. My guess about this is that either they don't get created at all or that they get deleted by `undo_ftplugin`. I don't know which of the 2. And btw, even when I had `fzf-lua` installed at the first half of the video, when I opened the second buffer from Neotree instead of fzf-lua the LSP mappings were there.

**Reply by chronon** [reply]
Thanks for checking into this @ibhagwan and thanks @dpetka2001 for the explanation and video. All I can add is that the behavior seen in the video matches what I'm seeing as well.

**Reply by dpetka2001** [reply]
Just to clear any potential misunderstandings. I don't know why this happens. And I'm not blaming fzf-lua for the behavior that I've observed. If i knew what was actually happening and knew for certain that fzf-lua was the culprit, I would have created an issue instead. But because I'm not sure what's happening and it sounds reasonable to me that Neovim in combination with fzf-lua might be causing this, that's why I preferred to discuss this with you @ibhagwan, so that you can also give your insight on this.

The fact is that fzf-lua has a different behavior vs opening a file via Neotree or Telescope. No idea why.

**Reply by ibhagwan** [reply]
I'm still unable to reprpduce this with my config, this seems to be a combination of how these keys are defined in LazyVim, which I assume should be in the LSP `on_attach` callback, perhaos you can check that?

In my config I added the below to my LSP `on_attach`:
```lua
  print("on_attach", bufnr)
```

And also added the below to the first lines of `/usr/share/nvim/runtime/ftplugin/php.vim`:
```lua
lua print(vim.api.nvim_get_current_buf(),"did_ftplugin", vim.b.did_ftplugin)
```
![image](https://github.com/user-attachments/assets/593a1177-5b51-433a-8742-0ae663fca2a3)


When I open the first file using `vi src/RemoteFilesPlugin.php`, I get:
![image](https://github.com/user-attachments/assets/86b6b60c-4bdf-4867-8866-0145e9bb86ae)

And my custom mapping is also there:
![image](https://github.com/user-attachments/assets/92df79f2-287d-49e0-9007-854d010f8159)

When I open the second file with `fzf-lua`, I get this in `:messages`:
![image](https://github.com/user-attachments/assets/be015a2e-a65e-4ae4-9371-5fbd6bc9e4fe)

And my custom map `Test` is also there:
![image](https://github.com/user-attachments/assets/d21bc6dc-a203-4a8e-886f-7b2d91e65304)

You can see the ordering, first the ftplugin is called, then the LSP `on_attach` is called which sets the custom map.

**It seems that the function setting the LSP mappings in LazyVim is breaking or isn't being called** - that should be our debugging direction.

**Reply by ibhagwan** [reply]
@dpetka2001, can you start debugging this function and see if `on_attach` is getting called after opening a file from fzf-lua?

The issue is most likely here:
https://github.com/LazyVim/LazyVim/blob/2cf3d234bfc31a1a4eda282ddabc12987f762893/lua/lazyvim/util/lsp.lua#L26

**Reply by dpetka2001** [reply]
Thank you for your insight. I'll be out for the rest of the evening, but I'll try to look into it again tomorrow and try to figure out what's happening based on your latest insight. Thank you as always.

**Reply by dpetka2001** [reply]
@ibhagwan I just tried to dive into it tonight after all. Could you see my [latest answer in LazyVim discussions](https://github.com/LazyVim/LazyVim/discussions/4966#discussioncomment-11585439) and give your insight please? It seems that when opening a file from fzf-lua, the `LspAttach` event triggers when it's still terminal mode.

**Reply by dpetka2001** [reply, marked as answer]
@ibhagwan It has been solved after all. I want to really thank you once again for being patient and willing to look into this and provide the necessary insight and guidance, so I could find the culprit. In the end, I only managed to find out where things went wrong but not the why, but thankfully Folke took it over from there and came up with a proper solution.

Again, you are a great person and developer and really thank you, because every time I've interacted with you, you've been really helpful :smile: Greatly appreciated!!

**Reply by ibhagwan** [reply]
> @ibhagwan I just tried to dive into it tonight after all. Could you see my [latest answer in LazyVim discussions](https://github.com/LazyVim/LazyVim/discussions/4966#discussioncomment-11585439) and give your insight please? It seems that when opening a file from fzf-lua, the `LspAttach` event triggers when it's still terminal mode.

That because mode changes in vim are scheduled, that’s why it takes time to switch back to normal mode and the actions are still performed in terminal mode.

I see this got solved, very nice :)

**Reply by ibhagwan** [reply]
> Again, you are a great person and developer and really thank you, because every time I've interacted with you, you've been really helpful 😄 Greatly appreciated!!

Tysm for the kind words @dpetka2001 ❤️ 

Ty @folke for fixing this pesky issue!
